### DISCUSSION TITLE
RaisePropertyChanged won't change UI while using "DeepData"

### DISCUSSION THREAD

**Discussion by Tokeyman** [AvaloniaUI/Avalonia#14752] [Q&A] [answered]
background: AvaloniaUI 11 ReactiveUI
here is the code in axaml:
```xaml
<StackPanel>
	<TextBlock Text="{Binding Greeting}" />
	<TextBlock Text="{Binding Path=Data.Name}" />
	<Button Content="Test" Command="{Binding Test}"/>
</StackPanel>
```
two TextBlock binding to a string and a "DeepData".
one button to test.
Here is the code in ViewModel:
```c#
 public class DeepData     //assuming this is an EntityModel
 {
     public string Name { get; set; } = "aaa";

     public DeepData() { }
     public DeepData(string name)  {  this.Name = name;  }
 }


public class MainViewModel : ViewModelBase //ReactiveObject
{
    public string Greeting { get; set; } = "Welcome to Avalonia!";

    private DeepData _deepData;
    public DeepData DeepData { get => _deepData; set => this.RaiseAndSetIfChanged(ref _deepData, value); }

    public MainViewModel()
    {
        DeepData = new DeepData();
    }

    public void Test()
    {
        DeepData.Name = "test";                                     //not change UI
        this.RaisePropertyChanged(nameof(DeepData));
        DeepData = new DeepData("ccc");                       // change UI
        Greeting = "he";
        this.RaisePropertyChanged(nameof(Greeting));  //also change UI
    }
}
```
Same code will work fine in WPF.
did i miss something?

**Comment by IanRawley** [comment]
Your DeepData class doesn't implement INotifyPropertyChanged, so there's no way for the UI to know the name has changed.  The MainViewModel does, so that's why it changes when you change the actual DeepData object.

**Reply by BAndysc** [reply]
That's not what the author asks for tho. The author raises Property Changes on DeepData after setting the inner property and indeed this works in WPF whereas it doesn't in Avalonia. Is it because Avalonia checks whether the property is actually changed and WPF doesn't? Dunno. Should Avalonia work like WPF in this regard?

**Reply by IanRawley** [reply]
At a guess the compiler is interpreting "nameof(DeepData)" as the class rather than property.  What happens if the class and property name are distinct?

**Reply by BAndysc** [reply]
Hmm? nameof() is a compile time expression, it is just a string. Doesn't matter what the name is. Maybe I wasn't clear, it is about the semantics, if you invoke PropertyChanged on a property, even if the value didn't change, WPF updates all nested bindings, Avalonia doesn't. It doesn't necessarily has to be a bug, this is a design choice. I don't know if it was intended tho.

**Comment by timunie** [comment, marked as answer]
I think this is the answer: https://github.com/AvaloniaUI/Avalonia/issues/14453

**Reply by Tokeyman** [reply]
After chang my Code as below. Neted property binding could work.
``` c#
public class DeepData : ReactiveObject
{
    private string _name = "aaa";
    public string Name { get => _name; set => this.RaiseAndSetIfChanged(ref _name, value); }

    public DeepData() { }
    public DeepData(string name)
    {
        this.Name = name;
    }
}


public class MainViewModel : ViewModelBase
{
    public string Greeting { get; set; } = "Welcome to Avalonia!";

    private DeepData _deepData;
    public DeepData DeepData { get => _deepData; set => this.RaiseAndSetIfChanged(ref _deepData, value); }

    public MainViewModel()
    {
        DeepData = new DeepData();
    }

    public void Test()
    {
        DeepData.Name = "test";                                     //work fine.

    }

}
```

Change the property to be observable will work fine.
So. I have to change every property to be observable deep into every EntityModel. that a huge work. 
But here is another question: if an EntityModel does not support to edit (like a nuget package), the only way to implement RaisePropertyChange is to write my own "proxy class" and rewrite all the properties ? right?
I do not have a lot of experience using WPF. In my past project. I used MVVM pattern like this:
Model Layer is EntityModel simply getter and setter.
```c#
public class Model
{
     public string? Name { get; set;}
}
```
ViewModel would be like UI's backend and transfer/process data between Model and View

```c#
public class ViewModel:ViewModelBase
{
    private Model _model;
    public Model Model { get=>_model; set=>RaisePropertyChange(ref _model,value); }
    
    private string? otherProperty;
    public string? OtherProperty { get=>otherProperty; set=>RaisePropertyChange(ref otherProperty,value);}
    
    public void Command()
    {
        // do something change property values and fire PropertyChange.
         RaisePropertyChange(nameof(Model));
    }
}
```

View is binding ViewModel's property;

```xaml
<Label Content="{Binding Model.Name"/>
<Label Content="{Binding OtherProperty "/>
```

If I use Avalonia. the Model layer would not be so this clean. the Model class has to be inherited ReactiveObject and implement PropertyChange . Am i right?

**Reply by timunie** [reply]
It only needs INotifyPropertyChanged. Can also be own impl or MVVM Toolkit.

**Reply by BAndysc** [reply]
I am afraid we have diverged from the original question again. The question is: if the PropertyChanged event is raised on a property whose value hasn't really changed (in case of reference type like here - it is the same reference), should it cause nested binding to update? Because it does work like that in WPF, but does not in Avalonia

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/issues/14453

**Reply by BAndysc** [reply]
Ah okay, it is all clear now, thanks for the clarification, sorry for my confusion.
